src/vs/platform/actionWidget/browser/actionWidget.ts

```diff
@@ -0,0 +1,634 @@
+}
+
+export interface IActionMenuTemplateData {
+	readonly container: HTMLElement;
+	readonly icon: HTMLElement;
+	readonly text: HTMLElement;
+	readonly keybinding: KeybindingLabel;
+}
+
+export enum ActionListItemKind {
+	Action = 'action',
+	Header = 'header'
+}
+
+interface HeaderTemplateData {
+	readonly container: HTMLElement;
+	readonly text: HTMLElement;
+}
+
+export class HeaderRenderer<T extends ListMenuItem<T>> implements IListRenderer<T, HeaderTemplateData> {
+
+	get templateId(): string { return ActionListItemKind.Header; }
+
+	renderTemplate(container: HTMLElement): HeaderTemplateData {
+		container.classList.add('group-header');
+
+		const text = document.createElement('span');
+		container.append(text);
+
+		return { container, text };
+	}
+
+	renderElement(element: ListMenuItem<T>, _index: number, templateData: HeaderTemplateData): void {
+		templateData.text.textContent = element.group.title;
+	}
+
+	disposeTemplate(_templateData: HeaderTemplateData): void {
+		// noop
+	}
+}
+
+
+export class ActionItemRenderer<T extends ListMenuItem<T>> implements IListRenderer<T, IActionMenuTemplateData> {
+
+	get templateId(): string { return 'action'; }
+
+	constructor(@IKeybindingService private readonly _keybindingService: IKeybindingService,
+		private readonly _resolver?: any) {
+	}
+
+	renderTemplate(container: HTMLElement): IActionMenuTemplateData {
+		container.classList.add(this.templateId);
+
+		const icon = document.createElement('div');
+		icon.className = 'icon';
+		container.append(icon);
+
+		const text = document.createElement('span');
+		text.className = 'title';
+		container.append(text);
+
+		const keybinding = new KeybindingLabel(container, OS);
+
+		return { container, icon, text, keybinding };
+	}
+
+	renderElement(element: T, _index: number, data: IActionMenuTemplateData): void {
+		if (element.group.icon) {
+			data.icon.className = element.group.icon.codicon.classNames;
+			data.icon.style.color = element.group.icon.color ?? '';
+		} else {
+			data.icon.className = Codicon.lightBulb.classNames;
+			data.icon.style.color = 'var(--vscode-editorLightBulb-foreground)';
+		}
+		if (!element.item || !element.label) {
+			return;
+		}
+		data.text.textContent = stripNewlines(element.label);
+		let binding;
+		if (this._resolver) {
+			binding = this._resolver?.getResolver()((element.item as any).action);
+		}
+		data.keybinding.set(binding);
+		if (!binding) {
+			dom.hide(data.keybinding.element);
+		} else {
+			dom.show(data.keybinding.element);
+		}
+
+		if (element.disabled) {
+			data.container.title = element.label;
+			data.container.classList.add('option-disabled');
+		} else {
+			data.container.title = localize({ key: 'label', comment: ['placeholders are keybindings, e.g "F2 to Apply, Shift+F2 to Preview"'] }, "{0} to Apply, {1} to Preview", this._keybindingService.lookupKeybinding('acceptSelectedCodeAction')?.getLabel(), this._keybindingService.lookupKeybinding('previewSelectedCodeAction')?.getLabel());
```

Comment:
The keybinding labels could come back as nothing

Reply:
not sure when that would happen